crates/matrix-sdk/src/client/mod.rs

```diff
@@ -952,6 +952,78 @@ impl Client {
         self.login_with_token(token.as_str(), device_id, initial_device_display_name).await
     }
 
+    /// Login as an appservice-generated virtual user
+    ///
+    /// This function can be used to create an end-to-end-encryption capable
+    /// appservice virtual user.
+    ///
+    /// This [`Client`] needs to be an appservice client configured with the
+    /// as_token. The returned session needs to be passed to
+    /// [`Client::restore_login()`].
+    ///
+    /// # Example
+    /// ```no_run
+    /// # use matrix_sdk::Client;
+    /// # async fn run(appservice_client: Client) -> anyhow::Result<()> {
+    ///   let sess = appservice_client.appservice_session("some_user", None, None).await?;
+    ///
+    ///   // Persist the session
+    ///
+    ///   let client = Client::builder()
+    ///                       .appservice_mode()
+    ///                       .homeserver_url(appservice_client.homeserver().await)
+    ///                       .build()
+    ///                       .await?;
+    ///
+    ///   client.restore_login(sess).await?;
+    /// # anyhow::Ok(())
+    /// # }
+    /// ```
+    ///
+    /// # End-to-end Encryption
+    ///
+    /// In order to use end-to-end encryption you currently need to log in as a
+    /// virtual appservice user as above, and sync like you would a normal
+    /// client. This is because the current appservice event pushing API does
+    /// not support pushing device events to the appservice.
+    ///
+    /// # Arguments
+    ///
+    /// * `user` - The user that should be logged in to the homeserver.
+    ///
+    /// * `device_id` - A unique id that will be associated with this session.
+    ///   If not given the homeserver will create one. Can be an existing
+    ///   device_id from a previous login call. Note that this should be done
+    ///   only if the client also holds the encryption keys for this device.
+    #[instrument(skip(self, user))]
+    #[cfg(feature = "appservice")]
+    pub async fn appservice_session(
```

Comment:
Is there a reason that this method has to live in the `Client`? If I'm not mistaken it could also just live in the `AppService` struct?

Reply:
OTOH putting it here would mean that the appservice-sdk woudl no longer be compatible with the latest released matrix sdk

Reply:
I'd like to keep the appservice-related code out of the other SDK crates as much as possible. So, since we can move this method to the appservice crate, we should do so.